Nodejs/Product/Nodejs/Workspace/NpmActionProviderFactory.cs

```diff
@@ -0,0 +1,168 @@
+            {
+                using (var npmController = this.CreateController())
+                using (var npmWorker = new NpmWorker(npmController))
+                using (var manager = new NpmPackageInstallWindow(npmController, npmWorker))
+                {
+                    manager.ShowModal();
+                }
+
+                return new FileContextActionResult(true);
+            }
+        }
+
+        private class UpdateNpmPackagesAction : NpmPackagesAction
+        {
+            public UpdateNpmPackagesAction(string packageJsonPath, FileContext fileContext, FileContextActionProvider provider) :
+                base(packageJsonPath, fileContext, PkgCmdId.cmdidWorkSpaceNpmUpdate, provider)
+            {
+            }
+
+            public override async Task<IFileContextActionResult> ExecuteAsync(IProgress<IFileContextActionProgressUpdate> progress, CancellationToken cancellationToken)
+            {
+                using (var npmController = this.CreateController())
+                using (var commander = npmController.CreateNpmCommander())
+                {
+                    var result = await commander.UpdatePackagesAsync();
+                    return new FileContextActionResult(result);
+                }
+            }
+        }
+
+        private abstract class NpmPackagesAction : IFileContextAction, IVsCommandItem
+        {
+            public NpmPackagesAction(string packageJsonPath, FileContext source, uint commandId, FileContextActionProvider provider)
+            {
+                this.Source = source;
+                this.CommandId = commandId;
+                this.PackageJsonPath = packageJsonPath;
+                this.WriteLine = provider.outputPane.WriteLine;
+                this.JTF = provider.workspaceContext.JTF;
+            }
+
+            public INpmController CreateController()
+            {
+                var projectHome = Path.GetDirectoryName(this.PackageJsonPath);
+
+                var npmController = NpmControllerFactory.Create(
+                      projectHome,
+                      NodejsConstants.NpmCachePath);
+
+                npmController.ErrorLogged += this.WriteNpmOutput;
+                npmController.ExceptionLogged += this.WriteNpmException;
+                npmController.OutputLogged += this.WriteNpmOutput;
+
+                return npmController;
+            }
+
+            private void WriteNpmException(object sender, NpmExceptionEventArgs e)
+            {
+                this.WriteLine(ErrorHelper.GetExceptionDetailsText(e.Exception));
+            }
+
+            private void WriteNpmOutput(object sender, NpmLogEventArgs e)
+            {
+                if (!string.IsNullOrWhiteSpace(e.LogText))
+                {
+                    this.WriteLine(e.LogText.TrimEnd('\r', '\n'));
+                }
+            }
+
+            protected readonly Action<string> WriteLine;
```

Comment:
Personally, I would expose a method to subclasses.

Reply:
Since they're all private classes.. I'm ok with this.